packages/awslint/lib/rules/api.ts

```diff
@@ -3,143 +3,220 @@ import { ConstructReflection } from './construct';
 import { CoreTypes } from './core-types';
 import { Linter } from '../linter';
 
-const EXCLUDE_ANNOTATION_REF_VIA_INTERFACE = '[disable-awslint:ref-via-interface]';
+const EXCLUDE_ANNOTATION_REF_VIA_INTERFACE =
+  '[disable-awslint:ref-via-interface]';
+
+const EXCLUDE_ANNOTATION_REF_VIA_REF_INTERFACE =
+  '[disable-awslint:ref-via-ref-interface]';
 
 // lint all constructs that are not L1 resources
-export const apiLinter = new Linter(a => ConstructReflection
-  .findAllConstructs(a)
-  .filter(c => !CoreTypes.isCfnResource(c.classType)));
+export const apiLinter = new Linter((a) =>
+  ConstructReflection.findAllConstructs(a).filter(
+    (c) => !CoreTypes.isCfnResource(c.classType),
+  ),
+);
+
+apiLinter.add({
+  code: 'ref-via-ref-interface',
+  message:
+    'API should use reference interfaces and not construct interface (%s). ' +
+    `If this is intentional, add "${EXCLUDE_ANNOTATION_REF_VIA_REF_INTERFACE}" to element's jsdoc`,
+  eval: (e) => {
+    const cls = e.ctx.classType;
+    visitClass(cls, {
+      // Any interfaces should be `ICfnAbc`
+      assertInterfaceType: (type, docs, scope) => {
+        // Receiving a generic construct is allowed
+        if (type.fqn === 'constructs.IConstruct') {
+          return;
+        }
+
+        // allow exclusion of this rule
+        if (
+          docs.summary.includes(EXCLUDE_ANNOTATION_REF_VIA_REF_INTERFACE) ||
+          docs.remarks.includes(EXCLUDE_ANNOTATION_REF_VIA_REF_INTERFACE)
+        ) {
+          return;
+        }
+
+        e.assert(type.name.startsWith('ICfn'), scope, type.fqn);
+      },
+    });
+  },
+});
 
 apiLinter.add({
   code: 'ref-via-interface',
-  message: 'API should use interface and not the concrete class (%s). ' +
+  message:
+    'API should use interface and not the concrete class (%s). ' +
     `If this is intentional, add "${EXCLUDE_ANNOTATION_REF_VIA_INTERFACE}" to element's jsdoc`,
-  eval: e => {
+  eval: (e) => {
```

Comment:
This rule hasn't changed. I just refactored the code used to visit spec nodes to be reusable.